src/utils/k6/testRun.ts

```diff
@@ -0,0 +1,204 @@
+  InvalidConfig = 104,
+
+  // ExternalAbort indicates the test was aborted by an external signal
+  // (e.g. SIGINT, SIGTERM, etc.) and should be considered aborted rather
+  // than a failure.
+  ExternalAbort = 105,
+
+  // CannotStartRESTAPI indicates the k6's REST API server could not be started.
+  CannotStartRESTAPI = 106,
+
+  // ScriptException indicates an exception was thrown during the
+  // test script's execution.
+  ScriptException = 107,
+
+  // ScriptAborted indicates the script was aborted by a call to the
+  // k6 execution module's `test.abort()` function.
+  ScriptAborted = 108,
+
+  // GoPanic indicates the script was aborted by a panic in the Go runtime.
+  GoPanic = 109,
+
+  // MarkedAsFailed indicates that the test was marked as failed.
+  MarkedAsFailed = 110,
+}
+
+interface PassedTestResult {
+  passed: true
+}
+
+interface FailedTestResult {
+  passed: false
+}
+
+type TestResult = PassedTestResult | FailedTestResult
+
+export interface TestRunStartEvent {}
+
+export interface TestRunErrorEvent {
+  error: Error
+}
+
+export interface TestRunAbortEvent {}
+
+export interface TestRunDoneEvent {
+  result: TestResult
+  checks: Check[]
+}
+
+export interface TestRunLogEvent {
+  entry: LogEntry
+}
+
+interface TestRunEventMap {
+  start: TestRunStartEvent
+  error: TestRunErrorEvent
+  abort: TestRunAbortEvent
+  done: TestRunDoneEvent
+  log: TestRunLogEvent
+}
+
+export class TestRun extends EventEmitter<TestRunEventMap> {
+  #process: ChildProcessWithoutNullStreams
+
+  #checks: Check[] = []
+
+  constructor(process: ChildProcessWithoutNullStreams) {
+    super()
+
+    process.on('spawn', this.#handleStart)
+
+    process.on('error', this.#handleError)
+
+    process.on('close', this.#handleClose)
+
+    readline.createInterface(process.stdout).on('line', (line) => {
+      const checks = parseJsonAsSchema(line, CheckArraySchema)
+
+      if (checks.success) {
+        this.#checks.push(...checks.data)
+
+        return
+      }
+
+      const log = parseJsonAsSchema(line, LogEntrySchema)
+
+      if (!log.success) {
+        return
+      }
+
+      this.emit('log', { entry: log.data })
+    })
+
+    readline.createInterface(process.stderr).on('line', (line) => {
+      const log = parseJsonAsSchema(line, LogEntrySchema)
+
+      if (!log.success) {
+        return
+      }
+
+      this.emit('log', { entry: log.data })
+    })
+
+    this.#process = process
+  }
+
+  isRunning(): boolean {
+    return this.#process.pid != null && this.#process.exitCode === null
+  }
+
+  stop(): Promise<void> {
+    if (!this.isRunning()) {
+      return Promise.resolve()
+    }
+
+    return new Promise((resolve) => {
+      this.#process.once('close', resolve)
+
+      this.#process.kill()
+    })
+  }
+
+  #handleStart = () => {
+    this.emit('start', {})
+  }
+
+  #handleError = (error: Error) => {
+    this.emit('error', { error })
+  }
+
+  #handleClose = (code: number | null) => {
+    switch (code) {
```

Comment:
I've tried to map the exit codes that I thought made sense for our use case. Some things I've mapped because "why not?", others I've just ignored.